README.md
<img src="https://avatars.githubusercontent.com/u/40179672" width="75">

[![hack.d Lawrence McDaniel](https://img.shields.io/badge/hack.d-Lawrence%20McDaniel-orange.svg)](https://lawrencemcdaniel.com)
[![discuss.overhang.io](https://img.shields.io/static/v1?logo=discourse&label=Forums&style=flat-square&color=ff0080&message=discuss.overhang.io)](https://discuss.overhang.io)
[![docs.tutor.overhang.io](https://img.shields.io/static/v1?logo=readthedocs&label=Documentation&style=flat-square&color=blue&message=docs.tutor.overhang.io)](https://docs.tutor.overhang.io)<br/>
[![AWS](https://img.shields.io/badge/AWS-%23FF9900.svg?style=for-the-badge&logo=amazon-aws&logoColor=white)](https://aws.amazon.com/)
[![Docker](https://img.shields.io/badge/docker-%230db7ed.svg?style=for-the-badge&logo=docker&logoColor=white)](https://www.docker.com/)
[![Kubernetes](https://img.shields.io/badge/kubernetes-%23326ce5.svg?style=for-the-badge&logo=kubernetes&logoColor=white)](https://kubernetes.io/)

# tutor-plugin-build-mfe

**DEPRECATED**: tutor version v14.x and newer maintains a Docker build of the mfe plugin. Hereon, you should use Tutor's mfe build.

Github Action to use Tutor to build a Docker image of Open edX Micro Front-end (MFE) framework, and upload to an AWS Elastic Container Registry repository.

## About this plugin

The "Micro Frontend base plugin for Tutor" plugin is authored by overerhangio: [https://github.com/overhangio/tutor-mfe](https://github.com/overhangio/tutor-mfe).

This plugin makes it possible to easily add micro frontend (MFE) applications on top of an Open edX platform that runs with Tutor. To learn more about MFEs, please check the official Open edX documentation.

In addition, this plugin comes with a few MFEs which are enabled by default:

- Account
- Gradebook
- Learning
- Profile

## Usage

```yaml
name: Example workflow

on: workflow_dispatch

jobs:
  build:
    runs-on: ubuntu-latest

    steps:
      # required antecedent
      - uses: actions/checkout@v3.2.0

      # required antecedent
      - name: Configure AWS credentials
        uses: aws-actions/configure-aws-credentials@v1.7.0
        with:
          aws-access-key-id: ${{ secrets.THE_NAME_OF_YOUR_AWS_ACCESS_KEY_ID }}
          aws-secret-access-key: ${{ secrets.THE_NAME_OF_YOUR_AWS_SECRET_ACCESS_KEY }}
          aws-region: us-east-2

      # install and configure tutor and kubectl
      - name: Configure Github workflow environment
        uses: openedx-actions/tutor-k8s-init@v1.0.4

      # This action.
      # Note that aws-ecr-repo is optional. The default value is openedx_mfe
      - name: Build the image and upload to AWS ECR
        uses: openedx-actions/tutor-plugin-build-mfe@0.1.6
        with:
          aws-ecr-repository: openedx_mfe
          subdomain: 'www'
          domain-name: 'example.com'
          site-name: 'School of Rock'
```


CHANGELOG.md
# CHANGE LOG

## Version 0.1.7 (2023-06-27)

- version bumps

## Version 0.1.6 (2022-12-23)

* refactor for Tutor version 15

## Version 0.1.5 (2022-06-07)

* add MFE suffix to env variables

## Version 0.1.3 (2022-06-07)

* switch to $GITHUB_ENV

## Version 0.0.1 (2022-05-28)

**Experimental. Do not use in production.**

* Initial Git import


scripts/install-mfe.sh
#!/bin/bash
#---------------------------------------------------------
#
# - Download and tutor-install the MFE plugin.
# - Set the URL of the Docker image so that we can push to a
#   public repository in a later step.
# - build the environment files
#
#---------------------------------------------------------


pip install tutor-mfe
tutor plugins enable mfe
tutor config save --set MFE_DOCKER_IMAGE=${ECR_REGISTRY}/${ECR_REPOSITORY}:${REPOSITORY_TAG}

# build the production environment file
cat <<EOT >> "$(tutor config printroot)/env/plugins/mfe/build/mfe/env/production"
ACCESS_TOKEN_COOKIE_NAME='edx-jwt-cookie-header-payload'
BASE_URL='https://apps.${SUBDOMAIN}.${DOMAIN_NAME}'
CSRF_TOKEN_API_PATH='/csrf/api/v1/token'
CREDENTIALS_BASE_URL='https://credentials.${SUBDOMAIN}.${DOMAIN_NAME}'
DISCOVERY_API_BASE_URL='https://discovery.${SUBDOMAIN}.${DOMAIN_NAME}'
ECOMMERCE_BASE_URL='https://ecommerce.${SUBDOMAIN}.${DOMAIN_NAME}'
ENABLE_NEW_RELIC='false'
FAVICON_URL='https://cdn.${SUBDOMAIN}.${DOMAIN_NAME}/favicon.ico'
LANGUAGE_PREFERENCE_COOKIE_NAME='openedx-language-preference'
LMS_BASE_URL='https://${SUBDOMAIN}.${DOMAIN_NAME}'
LOGIN_URL='https://${SUBDOMAIN}.${DOMAIN_NAME}/login'
LOGO_URL='https://cdn.${SUBDOMAIN}.${DOMAIN_NAME}/logo.png'
LOGO_TRADEMARK_URL='https://cdn.${SUBDOMAIN}.${DOMAIN_NAME}/logo.png'
LOGO_WHITE_URL='https://cdn.${SUBDOMAIN}.${DOMAIN_NAME}/logo.png'
LOGOUT_URL='https://${SUBDOMAIN}.${DOMAIN_NAME}/logout'
MARKETING_SITE_BASE_URL='https://${SUBDOMAIN}.${DOMAIN_NAME}'
NODE_ENV=production
PUBLISHER_BASE_URL=''
REFRESH_ACCESS_TOKEN_ENDPOINT='https://${SUBDOMAIN}.${DOMAIN_NAME}/login_refresh'
SEGMENT_KEY=''
SITE_NAME='${SITE_NAME}'
STUDIO_BASE_URL='https://studio.${SUBDOMAIN}.${DOMAIN_NAME}'
USER_INFO_COOKIE_NAME='user-info'
EOT

# build the development environment file
cat <<EOT >> "$(tutor config printroot)/env/plugins/mfe/build/mfe/env/development"
DISCOVERY_API_BASE_URL=''
LMS_BASE_URL='http://${SUBDOMAIN}:8000'
LOGIN_URL='http://${SUBDOMAIN}:8000/login'
LOGO_URL='http://${SUBDOMAIN}:8000/theming/asset/images/logo.png'
LOGO_TRADEMARK_URL='http://${SUBDOMAIN}:8000/theming/asset/images/logo.png'
LOGOUT_URL='http://${SUBDOMAIN}:8000/logout'
MARKETING_SITE_BASE_URL='http://${SUBDOMAIN}:8000'
NODE_ENV=development
REFRESH_ACCESS_TOKEN_ENDPOINT='http://${SUBDOMAIN}:8000/login_refresh'
STUDIO_BASE_URL='http://studio.${SUBDOMAIN}:8001'
EOT


scripts/install-plugin.sh
#!/bin/bash
#---------------------------------------------------------
#
# - Download and tutor-install the MFE plugin.
# - Set the URL of the Docker image so that we can push to a
#   public repository in a later step.
# - build the environment files
#
#---------------------------------------------------------


pip install tutor-mfe
tutor plugins enable mfe
tutor config save --set MFE_DOCKER_IMAGE=${AWS_ECR_REGISTRY_MFE}/${AWS_ECR_REPOSITORY_MFE}:${REPOSITORY_TAG_MFE}

# build the production environment file
cat <<EOT >> "$(tutor config printroot)/env/plugins/mfe/build/mfe/env/production"
ACCESS_TOKEN_COOKIE_NAME='edx-jwt-cookie-header-payload'
BASE_URL='https://apps.${SUBDOMAIN}.${DOMAIN_NAME}'
CSRF_TOKEN_API_PATH='/csrf/api/v1/token'
CREDENTIALS_BASE_URL='https://credentials.${SUBDOMAIN}.${DOMAIN_NAME}'
DISCOVERY_API_BASE_URL='https://discovery.${SUBDOMAIN}.${DOMAIN_NAME}'
ECOMMERCE_BASE_URL='https://ecommerce.${SUBDOMAIN}.${DOMAIN_NAME}'
ENABLE_NEW_RELIC='false'
FAVICON_URL='https://cdn.${SUBDOMAIN}.${DOMAIN_NAME}/favicon.ico'
LANGUAGE_PREFERENCE_COOKIE_NAME='openedx-language-preference'
LMS_BASE_URL='https://${SUBDOMAIN}.${DOMAIN_NAME}'
LOGIN_URL='https://${SUBDOMAIN}.${DOMAIN_NAME}/login'
LOGO_URL='https://cdn.${SUBDOMAIN}.${DOMAIN_NAME}/logo.png'
LOGO_TRADEMARK_URL='https://cdn.${SUBDOMAIN}.${DOMAIN_NAME}/logo.png'
LOGO_WHITE_URL='https://cdn.${SUBDOMAIN}.${DOMAIN_NAME}/logo.png'
LOGOUT_URL='https://${SUBDOMAIN}.${DOMAIN_NAME}/logout'
MARKETING_SITE_BASE_URL='https://${SUBDOMAIN}.${DOMAIN_NAME}'
NODE_ENV=production
PUBLISHER_BASE_URL=''
REFRESH_ACCESS_TOKEN_ENDPOINT='https://${SUBDOMAIN}.${DOMAIN_NAME}/login_refresh'
SEGMENT_KEY=''
SITE_NAME='${SITE_NAME}'
STUDIO_BASE_URL='https://studio.${SUBDOMAIN}.${DOMAIN_NAME}'
USER_INFO_COOKIE_NAME='user-info'
EOT

# build the development environment file
cat <<EOT >> "$(tutor config printroot)/env/plugins/mfe/build/mfe/env/development"
DISCOVERY_API_BASE_URL=''
LMS_BASE_URL='http://${SUBDOMAIN}:8000'
LOGIN_URL='http://${SUBDOMAIN}:8000/login'
LOGO_URL='http://${SUBDOMAIN}:8000/theming/asset/images/logo.png'
LOGO_TRADEMARK_URL='http://${SUBDOMAIN}:8000/theming/asset/images/logo.png'
LOGOUT_URL='http://${SUBDOMAIN}:8000/logout'
MARKETING_SITE_BASE_URL='http://${SUBDOMAIN}:8000'
NODE_ENV=development
REFRESH_ACCESS_TOKEN_ENDPOINT='http://${SUBDOMAIN}:8000/login_refresh'
STUDIO_BASE_URL='http://studio.${SUBDOMAIN}:8001'
EOT
